### PR TITLE
[Tensor] Scaling functions for `Tensor`

### PR DESCRIPTION
**Pull request by skykongkong8**
- For mixed precision training, we must support loss scaling to preserve round-off accuracy from half-precision.
- Scale factor for such training is generally a model hyperparameter, but here we gave it as a Tensor parameter for more subtle handling in lower level. Refer to each commit for propsed changes.
- Loss Scaler, which will govern all scaling situations for Tensor, Weight, Optimizer will be propsed in the near future.
- Add unittest accordingly.

For easier review, followings are the functions I revised to convey/consider `scale_factors_fp32` :
* `sum_by_batch`
* `sum`
* `multiply`
* `divide`
* `add`
* `dot`
* `transpose`
* `clone`
* `rotate_180`
* `scale`
* `descale`

### CODE REVIEW

File: nntrainer/tensor/blas_interface.cpp [hunk 1]
```diff
@@ -235,22 +235,24 @@ static void scopy_int8_to_fp16(const unsigned int N, const uint8_t *X,
 }
 
 static void ewvm_FP16(const unsigned int N, const _FP16 *X, const _FP16 *Y,
-                      _FP16 *Z) {
+                      _FP16 *Z, const float alpha) {
 #if (defined USE__FP16 && USE_NEON)
```
Comment (myungjoo) [inline nntrainer/tensor/blas_interface.cpp:239]: This preprocessor expression is not recommended.

https://stackoverflow.com/questions/1312132/boolean-in-ifdef-is-ifdef-a-b-the-same-as-if-defineda-definedb

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp:239]: Then, as I presume:
`defined USE__FP16 && USE_NEON` to `defined USE__FP16 && define USE_NEON` ?

Comment (djeong20) [reply nntrainer/tensor/blas_interface.cpp]: what's the difference b/w `USE__FP16` and `USE_NEON` in the first place?

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: First, my apologies for lack of understanding with those... but as far as I am concerned:

it is to differentiate fp16-enabled true/false with android/non-android build
neon is only valid for ARM.

Comment (djeong20) [reply nntrainer/tensor/blas_interface.cpp]: `USE_NEON` is currently enabled for both Android/non-android in `meson.build`. maybe we should fix this.

```
if get_option('enable-fp16')
   arch = target_machine.cpu_family()
   if get_option('platform') == 'android'
     add_project_arguments('-mfp16-format=ieee', language: ['c', 'cpp'])
     extra_defines += '-DENABLE_FP16=1'
     extra_defines += '-DUSE__FP16=1'
     extra_defines += '-DUSE_NEON=1'
   elif arch == 'aarch64'
     add_project_arguments('-mfp16-format=ieee', language: ['c', 'cpp'])
     extra_defines += '-DENABLE_FP16=1'
     extra_defines += '-DUSE__FP16=1'
     extra_defines += '-DUSE_NEON=1'
   else
     has_avx512fp16 = cc.has_argument('-mavx512fp16')
     if (has_avx512fp16)
       # add_project_arguments(['-mavx512fp16'], language: ['c','cpp'])
       message ('Float16 for x86_64 enabled. Modern gcc-x64 genrally supports float16 with _Float16. -mavx512fp16 added for hardware acceleration')
       extra_defines += '-DENABLE_FP16=1'
     else
       warning ('Float16 for x86_64 enabled. However, software emulation is applied for fp16, making it slower and inconsistent. Use GCC 12+ for AVX512 FP16 support. This build will probably fail unless you bring a compiler that supports fp16 for x64.')
     endif
   endif  
endif
```

Comment (djeong20) [reply nntrainer/tensor/blas_interface.cpp]: Also, I think we can replace `defined USE__FP16 && defined USE_NEON` to just `#if USE_NEON` since `USE__FP16` is inclusive

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: First of all, `aarch64` also supports neon since it is 64bit ARM I guess?
Second of all, we still need it because neon also do work for float32.

For example,
1. fp16  / fp32 with no neon -> raw BLAS function or something else ( cBLAS / OpenCL, ... )
2. fp16 / fp32 with neon -> neon BLAS function

sometimes we might want to build with option such as: using neon but not fp16 (though it is not a typical case for now)

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: I mean we are making more options like:
fp32-exclusive / fp16fp32 both  || neon / non-neon

Comment (djeong20) [reply nntrainer/tensor/blas_interface.cpp]: thanks for clarifying! to summarize, would things be like this?

- [x] USE_NEON=1 & USE__FP16=1: fp16/fp32 with neon BLAS
- [ ] (WTD) USE_NEON=1 & USE__FP16=0: fp32 only with neon BLAS 
- [ ] (WTD) USE_NEON=0 & USE__FP16=1: fp16/fp32 only with cBlas, openCL
- [ ] (WTD) USE_NEON=0 & USE__FP16=0: fp32 only with cBlas, openCL

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: Happy to help :) That is correct.

Comment (myungjoo) [reply nntrainer/tensor/blas_interface.cpp]: Change ```#if defined KEYWORD``` --> ```#if defined (KEYWORD)```.
And, for ```#if defined X && defined Y```, use ```if defined(X) && defined(Y)```, for the readability and compatibility with compilers.

Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: Thank you for your clear guide!
Applied suggested changes.

File: nntrainer/tensor/blas_interface.cpp [hunk 2]
```diff
@@ -182,7 +182,7 @@ static void scopy_float16_to_float32(const unsigned int N, const _FP16 *X,
   unsigned int incy = abs(incY);
   unsigned int incx = abs(incX);
 
-#if (defined USE__FP16 && USE_NEON)
+#if (defined(USE__FP16) && USE_NEON)
```
Comment (myungjoo) [inline nntrainer/tensor/blas_interface.cpp:185]: Because you've added ```-DUSE_NEON=1``` in meson.build, it should work, but in case you didn't do so in Android build, (I don't see -DUSE_NEON in jni directory) please correct this, too.

**Review by myungjoo — COMMENTED** [review]
Other than the one line not corrected, it looks good.

File: nntrainer/tensor/blas_interface.cpp (hunk 2 above)
Comment (skykongkong8) [reply nntrainer/tensor/blas_interface.cpp]: Apparently I missed it..
Fixed.

**Review by djeong20 — APPROVED** [review]
LGTM

**Review by djeong20 — COMMENTED** [review]
maybe we could use negative test cases as well? (e.g. try scaling FP32 tensor)

**Comment by skykongkong8** [comment]
> maybe we could use negative test cases as well? (e.g. try scaling FP32 tensor)

That would be meaningful, but I think it would be much  more efficient to work after applying your local work.
If we are able to separate Tensor files by type, it is quite straight-forward to do so I guess.
Or maybe you can temporally modify `Tensor::scale()` to check its `tensor_type` and throw `invalid_argument` to observe similar effect.

**Comment by jijoongmoon** [comment]
@djeong20 , I think we are also considering blas operation when we restructure the tensor for multi-type 
```
Tensor --> FloatTensor --> fp32blas --> naive
                                    --> openBLAS
       --> HalfTensor  --> fp16blas  --> naive
                                     --> NEON
```

**Review by Hyoungsukim — APPROVED** [review]
lgtm

File: nntrainer/tensor/blas_interface.cpp [hunk 3]
```diff
@@ -73,14 +73,14 @@
 
 namespace nntrainer {
 
-#ifdef ENABLE_FP16
+#if defined(ENABLE_FP16)
```
Comment (SeoHyungjun) [inline nntrainer/tensor/blas_interface.cpp:76]: Wouldn't ifdef be more concise if it only has one argument?

**Review by SeoHyungjun — APPROVED** [review]

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by myungjoo — APPROVED** [review]
Good to go. Please rebase and move on.

**Comment by skykongkong8** [comment]
This PR is deprecated. Close.
